File: src/lib/projection/engine/arithmetic.ts
type TokenKind = "number" | "identifier" | "plus" | "minus" | "star" | "slash" | "lparen" | "rparen" | "abs" | "rate" | "eof";

interface Token {
  kind: TokenKind;
  lexeme: string;
  value?: number;
  offset: number;
}

interface Context {
  postingAmounts: Map<string, number>;
  accountBalances: Record<string, number>;
  rate: number;
}

class Lexer {
  private pos = 0;

  constructor(private input: string) {}

  next(): Token {
    this.skipWhitespace();

    if (this.pos >= this.input.length) {
      return { kind: "eof", lexeme: "", offset: this.pos };
    }

    const ch = this.input[this.pos];

    if (ch >= "0" && ch <= "9") {
      return this.lexNumber();
    }

    if ((ch >= "a" && ch <= "z") || (ch >= "A" && ch <= "Z") || ch === "_") {
      return this.lexIdent();
    }

    const start = this.pos;
    this.pos += 1;

    switch (ch) {
      case "+": return { kind: "plus", lexeme: "+", offset: start };
      case "-": return { kind: "minus", lexeme: "-", offset: start };
      case "*": return { kind: "star", lexeme: "*", offset: start };
      case "/": return { kind: "slash", lexeme: "/", offset: start };
      case "(": return { kind: "lparen", lexeme: "(", offset: start };
      case ")": return { kind: "rparen", lexeme: ")", offset: start };
      default:
        throw new ParseError(`Unexpected character '${ch}'`, start);
    }
  }

  private skipWhitespace(): void {
    while (this.pos < this.input.length && this.input[this.pos] === " ") {
      this.pos += 1;
    }
  }

  private lexNumber(): Token {
    const start = this.pos;
    while (this.pos < this.input.length && this.input[this.pos] >= "0" && this.input[this.pos] <= "9") {
      this.pos += 1;
    }
    if (this.pos < this.input.length && this.input[this.pos] === ".") {
      this.pos += 1;
      while (this.pos < this.input.length && this.input[this.pos] >= "0" && this.input[this.pos] <= "9") {
        this.pos += 1;
      }
    }
    const lexeme = this.input.slice(start, this.pos);
    const value = Number(lexeme);
    if (!Number.isFinite(value)) {
      throw new ParseError(`Invalid number '${lexeme}'`, start);
    }
    return { kind: "number", lexeme, value, offset: start };
  }

  private lexIdent(): Token {
    const start = this.pos;
    while (
      this.pos < this.input.length &&
      ((this.input[this.pos] >= "a" && this.input[this.pos] <= "z") ||
        (this.input[this.pos] >= "A" && this.input[this.pos] <= "Z") ||
        (this.input[this.pos] >= "0" && this.input[this.pos] <= "9") ||
        this.input[this.pos] === "_")
    ) {
      this.pos += 1;
    }
    const lexeme = this.input.slice(start, this.pos);
    if (lexeme === "abs") {
      return { kind: "abs", lexeme, offset: start };
    }
    if (lexeme === "rate") {
      return { kind: "rate", lexeme, offset: start };
    }
    return { kind: "identifier", lexeme, offset: start };
  }
}

export class ParseError extends Error {
  constructor(message: string, public offset: number) {
    super(`${message} at position ${offset}`);
    this.name = "ParseError";
  }
}

class Parser {
  private tokens: Token[] = [];
  private pos = 0;

  parse(input: string): (context: Context) => number {
    const lexer = new Lexer(input);
    while (true) {
      const token = lexer.next();
      this.tokens.push(token);
      if (token.kind === "eof") {
        break;
      }
    }
    return this.expr();
  }

  private peek(): Token {
    return this.tokens[this.pos] ?? this.tokens[this.tokens.length - 1];
  }

  private advance(): Token {
    const token = this.peek();
    if (this.pos < this.tokens.length) {
      this.pos += 1;
    }
    return token;
  }

  private check(kind: TokenKind): boolean {
    return this.peek().kind === kind;
  }

  private match(kind: TokenKind): Token | null {
    if (this.check(kind)) {
      return this.advance();
    }
    return null;
  }

  private consume(kind: TokenKind, message: string): Token {
    if (this.check(kind)) {
      return this.advance();
    }
    throw new ParseError(message, this.peek().offset);
  }

  private expr(): (context: Context) => number {
    let node = this.term();

    while (this.check("plus") || this.check("minus")) {
      const op = this.advance();
      const right = this.term();
      const left = node;
      node = op.kind === "plus"
        ? (ctx) => left(ctx) + right(ctx)
        : (ctx) => left(ctx) - right(ctx);
    }

    return node;
  }

  private term(): (context: Context) => number {
    let node = this.unary();

    while (this.check("star") || this.check("slash")) {
      const op = this.advance();
      const right = this.unary();
      const left = node;
      node = op.kind === "star"
        ? (ctx) => left(ctx) * right(ctx)
        : (ctx) => {
            const divisor = right(ctx);
            if (divisor === 0) {
              throw new EvalError("Division by zero");
            }
            return left(ctx) / divisor;
          };
    }

    return node;
  }

  private unary(): (context: Context) => number {
    if (this.match("minus")) {
      const operand = this.unary();
      return (ctx) => -operand(ctx);
    }

    return this.call();
  }

  private call(): (context: Context) => number {
    if (this.match("abs")) {
      this.consume("lparen", "Expected '(' after 'abs'");
      const arg = this.expr();
      this.consume("rparen", "Expected ')' after abs argument");
      return (ctx) => Math.abs(arg(ctx));
    }

    return this.primary();
  }

  private primary(): (context: Context) => number {
    if (this.match("lparen")) {
      const node = this.expr();
      this.consume("rparen", "Expected ')'");
      return node;
    }

    if (this.check("number")) {
      const token = this.advance();
      const value = token.value!;
      return () => value;
    }

    if (this.check("rate")) {
      this.advance();
      return (ctx) => ctx.rate;
    }

    if (this.check("identifier")) {
      const token = this.advance();
      const name = token.lexeme;
      return (ctx) => {
        const postingVal = ctx.postingAmounts.get(name);
        if (postingVal !== undefined) {
          return postingVal;
        }
        const accountVal = ctx.accountBalances[name];
        if (accountVal !== undefined) {
          return accountVal;
        }
        return 0;
      };
    }

    throw new ParseError(`Expected number, identifier, or '('`, this.peek().offset);
  }
}

export class EvalError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "EvalError";
  }
}

export function parseArithmetic(input: string): (context: Context) => number {
  const parser = new Parser();
  return parser.parse(input);
}

export function evaluateArithmetic(input: string, context: Context): number {
  return parseArithmetic(input)(context);
}

File: backend/internal/domain/arithmetic.go
package domain

import (
	"fmt"
	"math"
	"strconv"
	"sync"
)

// Arithmetic expression language ported from simulation/arithmetic.ts.
// Grammar: expr := term (('+'|'-') term)* ; term := unary (('*'|'/') unary)* ;
// unary := '-' unary | call ; call := 'abs' '(' expr ')' | primary ;
// primary := number | 'rate' | identifier | '(' expr ')'.

type tokenKind int

const (
	tokNumber tokenKind = iota
	tokIdentifier
	tokPlus
	tokMinus
	tokStar
	tokSlash
	tokLParen
	tokRParen
	tokAbs
	tokRate
	tokEOF
)

type token struct {
	kind   tokenKind
	lexeme string
	value  float64
	offset int
}

// ParseError reports a lexing/parsing failure with the input offset.
type ParseError struct {
	Message string
	Offset  int
}

func (e *ParseError) Error() string {
	return fmt.Sprintf("%s at position %d", e.Message, e.Offset)
}

// EvalError reports an evaluation-time failure (division by zero).
type EvalError struct{ Message string }

func (e *EvalError) Error() string { return e.Message }

type lexer struct {
	input string
	pos   int
}

func (l *lexer) next() (token, error) {
	for l.pos < len(l.input) && l.input[l.pos] == ' ' {
		l.pos++
	}
	if l.pos >= len(l.input) {
		return token{kind: tokEOF, offset: l.pos}, nil
	}
	ch := l.input[l.pos]
	switch {
	case ch >= '0' && ch <= '9':
		return l.lexNumber()
	case (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z') || ch == '_':
		return l.lexIdent()
	case ch == '+':
		l.pos++
		return token{kind: tokPlus, lexeme: "+", offset: l.pos - 1}, nil
	case ch == '-':
		l.pos++
		return token{kind: tokMinus, lexeme: "-", offset: l.pos - 1}, nil
	case ch == '*':
		l.pos++
		return token{kind: tokStar, lexeme: "*", offset: l.pos - 1}, nil
	case ch == '/':
		l.pos++
		return token{kind: tokSlash, lexeme: "/", offset: l.pos - 1}, nil
	case ch == '(':
		l.pos++
		return token{kind: tokLParen, lexeme: "(", offset: l.pos - 1}, nil
	case ch == ')':
		l.pos++
		return token{kind: tokRParen, lexeme: ")", offset: l.pos - 1}, nil
	default:
		return token{}, &ParseError{Message: fmt.Sprintf("Unexpected character '%c'", ch), Offset: l.pos}
	}
}

func (l *lexer) lexNumber() (token, error) {
	start := l.pos
	for l.pos < len(l.input) && l.input[l.pos] >= '0' && l.input[l.pos] <= '9' {
		l.pos++
	}
	if l.pos < len(l.input) && l.input[l.pos] == '.' {
		l.pos++
		for l.pos < len(l.input) && l.input[l.pos] >= '0' && l.input[l.pos] <= '9' {
			l.pos++
		}
	}
	lexeme := l.input[start:l.pos]
	value, err := strconv.ParseFloat(lexeme, 64)
	if err != nil || math.IsInf(value, 0) || math.IsNaN(value) {
		return token{}, &ParseError{Message: fmt.Sprintf("Invalid number '%s'", lexeme), Offset: start}
	}
	return token{kind: tokNumber, lexeme: lexeme, value: value, offset: start}, nil
}

func (l *lexer) lexIdent() (token, error) {
	start := l.pos
	for l.pos < len(l.input) {
		ch := l.input[l.pos]
		if (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z') || (ch >= '0' && ch <= '9') || ch == '_' {
			l.pos++
			continue
		}
		break
	}
	lexeme := l.input[start:l.pos]
	switch lexeme {
	case "abs":
		return token{kind: tokAbs, lexeme: lexeme, offset: start}, nil
	case "rate":
		return token{kind: tokRate, lexeme: lexeme, offset: start}, nil
	}
	return token{kind: tokIdentifier, lexeme: lexeme, offset: start}, nil
}

// ParsedArithmetic is a compiled expression plus its identifier requirements.
// Closures may panic with *EvalError (division by zero); Evaluate converts
// that into an error.
type ParsedArithmetic struct {
	evaluate     func(ctx map[string]float64) float64
	requirements []string
}

func (p ParsedArithmetic) Evaluate(ctx map[string]float64) (result float64, err error) {
	defer func() {
		if r := recover(); r != nil {
			if e, ok := r.(*EvalError); ok {
				result = 0
				err = e
				return
			}
			panic(r)
		}
	}()
	return p.evaluate(ctx), nil
}

func (p ParsedArithmetic) Requirements() []string { return p.requirements }

type parser struct {
	tokens       []token
	pos          int
	requirements map[string]bool
}

func (p *parser) peek() token {
	if p.pos < len(p.tokens) {
		return p.tokens[p.pos]
	}
	return p.tokens[len(p.tokens)-1]
}

func (p *parser) advance() token {
	t := p.peek()
	if p.pos < len(p.tokens) {
		p.pos++
	}
	return t
}

func (p *parser) check(kind tokenKind) bool { return p.peek().kind == kind }

func (p *parser) match(kind tokenKind) bool {
	if p.check(kind) {
		p.advance()
		return true
	}
	return false
}

func (p *parser) consume(kind tokenKind, message string) (token, error) {
	if p.check(kind) {
		return p.advance(), nil
	}
	return token{}, &ParseError{Message: message, Offset: p.peek().offset}
}

func (p *parser) expr() (func(map[string]float64) float64, error) {
	leftNode, err := p.term()
	if err != nil {
		return nil, err
	}
	node := leftNode
	for p.check(tokPlus) || p.check(tokMinus) {
		op := p.advance()
		right, err := p.term()
		if err != nil {
			return nil, err
		}
		left := node
		if op.kind == tokPlus {
			node = func(ctx map[string]float64) float64 { return left(ctx) + right(ctx) }
		} else {
			node = func(ctx map[string]float64) float64 { return left(ctx) - right(ctx) }
		}
	}
	return node, nil
}

func (p *parser) term() (func(map[string]float64) float64, error) {
	unaryNode, err := p.unary()
	if err != nil {
		return nil, err
	}
	node := unaryNode
	for p.check(tokStar) || p.check(tokSlash) {
		op := p.advance()
		right, err := p.unary()
		if err != nil {
			return nil, err
		}
		left := node
		if op.kind == tokStar {
			node = func(ctx map[string]float64) float64 { return left(ctx) * right(ctx) }
		} else {
			node = func(ctx map[string]float64) float64 {
				divisor := right(ctx)
				if divisor == 0 {
					panic(&EvalError{Message: "Division by zero"})
				}
				return left(ctx) / divisor
			}
		}
	}
	return node, nil
}

func (p *parser) unary() (func(map[string]float64) float64, error) {
	if p.match(tokMinus) {
		operand, err := p.unary()
		if err != nil {
			return nil, err
		}
		return func(ctx map[string]float64) float64 { return -operand(ctx) }, nil
	}
	return p.call()
}

func (p *parser) call() (func(map[string]float64) float64, error) {
	if p.match(tokAbs) {
		if _, err := p.consume(tokLParen, "Expected '(' after 'abs'"); err != nil {
			return nil, err
		}
		arg, err := p.expr()
		if err != nil {
			return nil, err
		}
		if _, err := p.consume(tokRParen, "Expected ')' after abs argument"); err != nil {
			return nil, err
		}
		return func(ctx map[string]float64) float64 { return math.Abs(arg(ctx)) }, nil
	}
	return p.primary()
}

func (p *parser) primary() (func(map[string]float64) float64, error) {
	if p.match(tokLParen) {
		node, err := p.expr()
		if err != nil {
			return nil, err
		}
		if _, err := p.consume(tokRParen, "Expected ')'"); err != nil {
			return nil, err
		}
		return node, nil
	}
	if p.check(tokNumber) {
		t := p.advance()
		value := t.value
		return func(map[string]float64) float64 { return value }, nil
	}
	if p.match(tokRate) {
		p.requirements["rate"] = true
		return func(ctx map[string]float64) float64 { return ctx["rate"] }, nil
	}
	if p.check(tokIdentifier) {
		t := p.advance()
		name := t.lexeme
		p.requirements[name] = true
		return func(ctx map[string]float64) float64 {
			v, ok := ctx[name]
			if !ok {
				return math.NaN() // matches TS undefined propagation
			}
			return v
		}, nil
	}
	return nil, &ParseError{Message: "Expected number, identifier, or '('", Offset: p.peek().offset}
}

var exprCache = sync.Map{}

// ParseArithmetic compiles an arithmetic expression. Compiled expressions are
// memoized; expressions are immutable so sharing is safe across goroutines.
func ParseArithmetic(input string) (ParsedArithmetic, error) {
	if cached, ok := exprCache.Load(input); ok {
		return cached.(ParsedArithmetic), nil
	}
	lexer := &lexer{input: input}
	var tokens []token
	for {
		t, err := lexer.next()
		if err != nil {
			return ParsedArithmetic{}, err
		}
		tokens = append(tokens, t)
		if t.kind == tokEOF {
			break
		}
	}
	p := &parser{tokens: tokens, requirements: map[string]bool{}}
	expr, err := p.expr()
	if err != nil {
		return ParsedArithmetic{}, err
	}
	if _, err := p.consume(tokEOF, "Unexpected trailing input"); err != nil {
		return ParsedArithmetic{}, err
	}
	reqs := make([]string, 0, len(p.requirements))
	for name := range p.requirements {
		reqs = append(reqs, name)
	}
	sortStrings(reqs)
	parsed := ParsedArithmetic{evaluate: expr, requirements: reqs}
	exprCache.Store(input, parsed)
	return parsed, nil
}

func sortStrings(values []string) {
	for i := 1; i < len(values); i++ {
		for j := i; j > 0 && values[j] < values[j-1]; j-- {
			values[j], values[j-1] = values[j-1], values[j]
		}
	}
}

// ArithmeticRequirements returns sorted identifiers referenced by input.
func ArithmeticRequirements(input string) ([]string, error) {
	parsed, err := ParseArithmetic(input)
	if err != nil {
		return nil, err
	}
	return parsed.Requirements(), nil
}

// EvaluateArithmetic evaluates input against a concrete numeric context,
// defaulting missing requirement identifiers to 0 (legacy TS behavior).
func EvaluateArithmetic(input string, ctx map[string]float64) (float64, error) {
	parsed, err := ParseArithmetic(input)
	if err != nil {
		return 0, err
	}
	values := make(map[string]float64, len(parsed.requirements)+len(ctx))
	for k, v := range ctx {
		values[k] = v
	}
	for _, req := range parsed.requirements {
		if _, ok := values[req]; !ok {
			values[req] = 0
		}
	}
	return parsed.Evaluate(values)
}
